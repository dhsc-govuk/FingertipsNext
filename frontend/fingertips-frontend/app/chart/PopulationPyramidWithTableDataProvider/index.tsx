import { GetHealthDataForAnIndicatorInequalitiesEnum } from '@/generated-sources/ft-api-client';
import {
  administratorIndicatorID,
  areaCodeForEngland,
  nhsIndicatorIdForPopulation,
} from '@/lib/chartHelpers/constants';
import { SearchParams, SearchStateParams } from '@/lib/searchStateManager';
import {
  allAreaTypes,
  HierarchyNameTypes,
} from '@/lib/areaFilterHelpers/areaType';

import { ApiClientFactory } from '@/lib/apiClient/apiClientFactory';
import { PopulationPyramidWithTable } from '@/components/organisms/PopulationPyramidWithTable';
import { getHealthDataForIndicator } from '@/lib/ViewsHelpers';
import { SearchServiceFactory } from '@/lib/search/searchServiceFactory';

const getPopulationData = (
  populationIndicatorID: number,
  areaCodesToRequest: string[]
) => {
  return getHealthDataForIndicator(
    ApiClientFactory.getIndicatorsApiClient(),
    populationIndicatorID,
    [
      {
        areaCodes: areaCodesToRequest,
        inequalities: [
          GetHealthDataForAnIndicatorInequalitiesEnum.Age,
          GetHealthDataForAnIndicatorInequalitiesEnum.Sex,
        ],
      },
    ]
  );
};

interface PyramidContextProviderProps {
  areaCodes: string[];
  searchState: SearchStateParams;
}

export const PopulationPyramidWithTableDataProvider = async ({
  areaCodes,
  searchState,
}: PyramidContextProviderProps) => {
  const {
    [SearchParams.GroupSelected]: groupAreaSelected,
    [SearchParams.AreaTypeSelected]: areaTypeSelected,
  } = searchState;

  const areaCodesToRequest = [...areaCodes];
  if (!areaCodesToRequest.includes(areaCodeForEngland)) {
    areaCodesToRequest.push(areaCodeForEngland);
  }
  if (groupAreaSelected && groupAreaSelected != areaCodeForEngland) {
    areaCodesToRequest.push(groupAreaSelected);
  }

  const hierarchyName = allAreaTypes.find(
    (areaType) => areaType.key === areaTypeSelected
  );

  const populationIndicatorID =
    hierarchyName?.hierarchyName === HierarchyNameTypes.NHS
      ? nhsIndicatorIdForPopulation
      : administratorIndicatorID;

  const getPopulationIndicatorMetadata = () => {
    return SearchServiceFactory.getIndicatorSearchService().getIndicator(
      populationIndicatorID.toString()
    );
  };

  const { populationData, populationMetadata } = await (async () => {
    if (!populationIndicatorID) {
      return {
        populationData: undefined,
        populationMetadata: undefined,
      };
    }

    const [populationData, populationIndicatorMetadata] = await Promise.all([
      await getPopulationData(populationIndicatorID, areaCodesToRequest),
      await getPopulationIndicatorMetadata(),
    ]);

    return { populationData, populationMetadata: populationIndicatorMetadata };
  })();

  return (
    <PopulationPyramidWithTable
      healthDataForAreas={populationData?.areaHealthData ?? []}
      searchState={searchState}
      xAxisTitle="Age"
      yAxisTitle="Percentage of total population"
      dataSource={populationMetadata?.dataSource}
    />
  );
};
