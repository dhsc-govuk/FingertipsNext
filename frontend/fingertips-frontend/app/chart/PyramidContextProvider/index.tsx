import {
  GetHealthDataForAnIndicatorInequalitiesEnum,
  IndicatorWithHealthDataForArea,
} from '@/generated-sources/ft-api-client';
import { areaCodeForEngland } from '@/lib/chartHelpers/constants';
import {
  SearchParams,
  SearchStateManager,
  SearchStateParams,
} from '@/lib/searchStateManager';
import { HierarchyNameTypes } from '@/lib/areaFilterHelpers/areaType';

import {
  API_CACHE_CONFIG,
  ApiClientFactory,
} from '@/lib/apiClient/apiClientFactory';
import { PopulationPyramidWithTable } from '@/components/organisms/PopulationPyramidWithTable';
const enum PopulationIndicatorIdsTypes {
  ADMINISTRATIVE = 92708,
  NHS = 337,
}

interface PyramidContextProviderProps {
  areaCodes: string[];
  searchState: SearchStateParams;
}

export const PyramidContextProvider = async ({
  areaCodes,
  searchState,
}: PyramidContextProviderProps) => {
  const stateManager = SearchStateManager.initialise(searchState);

  const { [SearchParams.GroupAreaSelected]: groupAreaSelected } =
    stateManager.getSearchState();

  const areaCodesToRequest = (() => {
    const areaCodesToRequest = [...areaCodes];
    if (!areaCodesToRequest.includes(areaCodeForEngland)) {
      areaCodesToRequest.push(areaCodeForEngland);
    }
    if (groupAreaSelected && groupAreaSelected != areaCodeForEngland) {
      areaCodesToRequest.push(groupAreaSelected);
    }
    return areaCodesToRequest;
  })();

  const areasApi = ApiClientFactory.getAreasApiClient();
  const indicatorApi = ApiClientFactory.getIndicatorsApiClient();
  const populationDataForArea: IndicatorWithHealthDataForArea | undefined =
    await (async () => {
      try {
        const populationIndicatorID: number = await (async (
          areaCode: string
        ) => {
          const area = await areasApi.getArea({ areaCode: areaCode });
          if (area.areaType.hierarchyName == HierarchyNameTypes.NHS) {
            return PopulationIndicatorIdsTypes.NHS;
          }
          return PopulationIndicatorIdsTypes.ADMINISTRATIVE;
        })(areaCodesToRequest[0]);

        return await indicatorApi.getHealthDataForAnIndicator(
          {
            indicatorId: populationIndicatorID,
            areaCodes: areaCodesToRequest,
            inequalities: [
              GetHealthDataForAnIndicatorInequalitiesEnum.Age,
              GetHealthDataForAnIndicatorInequalitiesEnum.Sex,
            ],
          },
          API_CACHE_CONFIG
        );
      } catch (error) {
        console.error(
          'error getting population health indicator data for area',
          error
        );
      }
    })();

  return (
    <PopulationPyramidWithTable
      healthDataForAreas={populationDataForArea?.areaHealthData ?? []}
      searchState={searchState}
      xAxisTitle="Age"
      yAxisTitle="Population"
    />
  );
};
