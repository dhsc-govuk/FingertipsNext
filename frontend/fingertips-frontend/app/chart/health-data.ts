interface HealthCareData {
  areaCode: string;
  healthData: {
    year: number;
    count: number;
    value: number;
    lowerCi: number;
    upperCi: number;
  }[];
}
