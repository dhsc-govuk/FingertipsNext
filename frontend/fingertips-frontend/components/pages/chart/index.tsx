'use client';

import { LineChart } from '@/components/organisms/LineChart';
import { BackLink, H2 } from 'govuk-react';
import { LineChartTable } from '@/components/organisms/LineChartTable';
import { HealthDataForArea } from '@/generated-sources/ft-api-client';
import { SearchParams, SearchStateManager } from '@/lib/searchStateManager';
import { BarChart } from '@/components/organisms/BarChart';
import { PopulationPyramid } from '@/components/organisms/PopulationPyramid';
import { PopulationData } from '@/lib/chartHelpers/preparePopulationData';
import { ScatterChart } from '@/components/organisms/ScatterChart';
import { getEnglandDataForIndicatorIndex } from '@/lib/chartHelpers/chartHelpers';

type ChartProps = {
  data: HealthDataForArea[][];
  populationData?: PopulationData;
  searchedIndicator?: string;
  indicatorsSelected?: string[];
};

export function Chart({
  data,
  populationData,
  searchedIndicator,
  indicatorsSelected = [],
}: Readonly<ChartProps>) {
  const searchState = new SearchStateManager({
    [SearchParams.SearchedIndicator]: searchedIndicator,
    [SearchParams.IndicatorsSelected]: indicatorsSelected,
  });
  const backLinkPath = searchState.generatePath('/results');

  const englandBenchmarkData = getEnglandDataForIndicatorIndex(data, 0);

  return (
    <>
      <BackLink
        data-testid="chart-page-back-link"
        href={backLinkPath}
        aria-label="Go back to the previous page"
      />
      <H2>View data for selected indicators and areas</H2>
      {populationData ? (
        <>
          <PopulationPyramid
            data={populationData}
            populationPyramidTitle="Population INDICATOR for SELECTED area"
            xAxisTitle="Age"
            yAxisTitle="Percentage of total population"
            accessibilityLabel="A pyramid chart showing population data for SELECTED AREA"
          />
          <br />
        </>
      ) : null}
      {indicatorsSelected.length == 2 ? (
        <ScatterChart
          data={data}
          ScatterChartTitle="Compare indicators within the area group"
          yAxisTitle="y: Indicator 1 (value)"
          yAxisSubtitle="rate per information"
          xAxisTitle="x: Indicator 2 (value)"
          xAxisSubtitle="rate per information"
          accessibilityLabel="A scatter chart showing two indicators"
        ></ScatterChart>
      ) : null}
      <LineChart
        LineChartTitle="See how the indicator has changed over time"
        data={data[0]}
        xAxisTitle="Year"
        accessibilityLabel="A line chart showing healthcare data"
      />
      <br />
      <BarChart
        data={data[0]}
        yAxisTitle="Value"
        benchmarkLabel="England"
        benchmarkValue={800}
        accessibilityLabel="A bar chart showing healthcare data"
      />
      <br />
      <LineChartTable
        data={data[0][0]}
        englandBenchmarkData={englandBenchmarkData}
      ></LineChartTable>
    </>
  );
}
