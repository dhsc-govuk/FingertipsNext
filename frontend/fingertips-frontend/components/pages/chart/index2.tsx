'use client';

import { LineChart } from '@/components/linechartsOptions';
import { WeatherForecast } from '@/generated-sources/api-client';
import { H1 } from 'govuk-react';

type ChartProps = {
  data: WeatherForecast[];
};

export function Chart({ data }: Readonly<ChartProps>) {
  return (
    <>
      <H1>Line Chart</H1>
      {/*<StyledComponentsRegistry>*/}
      {/*  <Table*/}
      {/*    head={*/}
      {/*      <Table.Row>*/}
      {/*        <Table.CellHeader date>Date</Table.CellHeader>*/}
      {/*        <Table.CellHeader numeric>Temperature C</Table.CellHeader>*/}
      {/*        <Table.CellHeader numeric>Temperature F</Table.CellHeader>*/}
      {/*        <Table.CellHeader>Summary</Table.CellHeader>*/}
      {/*      </Table.Row>*/}
      {/*    }*/}
      {/*  >*/}
      {/*    {data.map((item) => (*/}
      {/*      <Table.Row key={`${item.date}-${item.temperatureC}`}>*/}
      {/*        <Table.Cell>{item.date}</Table.Cell>*/}
      {/*        <Table.Cell numeric>{item.temperatureC}</Table.Cell>*/}
      {/*        <Table.Cell numeric>{item.temperatureF}</Table.Cell>*/}
      {/*        <Table.Cell>{item.summary}</Table.Cell>*/}
      {/*      </Table.Row>*/}
      {/*    ))}*/}
      {/*  </Table>*/}
      {/*</StyledComponentsRegistry>*/}
      <div>
        <LineChart
          data={data}
          title="Weather Forecast"
          xAxisTitle="Date"
          yAxisTitle="Temperature (Â°C)"
        />
      </div>
    </>
  );
}
