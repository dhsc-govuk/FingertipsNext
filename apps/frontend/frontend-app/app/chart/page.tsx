import { connection } from 'next/server';
import {
  SearchParams,
  SearchStateManager,
  SearchStateParams,
} from '@/lib/searchStateManager';
import { getSelectedAreasDataByAreaType } from '@/lib/areaFilterHelpers/getSelectedAreasData';
import { AreaTypeKeys } from '@/lib/areaFilterHelpers/areaType';
import { ViewsContext } from '@/components/views/ViewsContext';
import { SearchServiceFactory } from '@/lib/search/searchServiceFactory';
import { getAreaFilterData } from '@/lib/areaFilterHelpers/getAreaFilterData';
import { ErrorPage } from '@/components/pages/error';

export default async function ChartPage(
  props: Readonly<{
    searchParams?: Promise<SearchStateParams>;
  }>
) {
  try {
    const searchParams = await props.searchParams;
    const stateManager = SearchStateManager.initialise(searchParams);
    const {
      [SearchParams.AreasSelected]: areaCodes,
      [SearchParams.IndicatorsSelected]: indicatorsSelected,
      [SearchParams.AreaTypeSelected]: areaTypeSelected,
    } = stateManager.getSearchState();

    const areasSelected = areaCodes ?? [];

    // We don't want to render this page statically
    await connection();

    const selectedIndicatorsData =
      indicatorsSelected && indicatorsSelected.length > 0
        ? (
            await Promise.all(
              indicatorsSelected.map((indicator) =>
                SearchServiceFactory.getIndicatorSearchService().getIndicator(
                  indicator
                )
              )
            )
          ).filter((indicatorDocument) => indicatorDocument !== undefined)
        : [];

    const selectedAreasData = await getSelectedAreasDataByAreaType(
      areasSelected,
      areaTypeSelected as AreaTypeKeys
    );

    const {
      availableAreaTypes,
      availableAreas,
      availableGroupTypes,
      availableGroups,
      updatedSearchState,
    } = await getAreaFilterData(
      stateManager.getSearchState(),
      selectedAreasData
    );

    if (updatedSearchState) {
      stateManager.setState(updatedSearchState);
    }

    return (
      <ViewsContext
        searchState={stateManager.getSearchState()}
        selectedAreasData={selectedAreasData}
        selectedIndicatorsData={selectedIndicatorsData}
        areaFilterData={{
          availableAreaTypes,
          availableGroupTypes,
          availableGroups,
          availableAreas,
        }}
      />
    );
  } catch (error) {
    console.log(`Error response received from call: ${error}`);
    return <ErrorPage />;
  }
}
