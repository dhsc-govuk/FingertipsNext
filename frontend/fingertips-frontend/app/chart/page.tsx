import { Chart } from '@/components/pages/chart';
import { connection } from 'next/server';
import { IndicatorsApi } from '@/generated-sources/ft-api-client';
import { getApiConfiguration } from '@/lib/getApiConfiguration';

export default async function ChartPage(
  props: Readonly<{
    searchParams?: Promise<{
      indicator?: string;
      indicatorsSelected?: string;
      areaCodes?: string;
    }>;
  }>
) {
  // We don't want to render this page statically
  await connection();

  const searchParams = await props.searchParams;
  const indicator = searchParams?.indicator ?? '';
  const indicatorsSelected = searchParams?.indicatorsSelected?.split(',') ?? [];
  const areaCodes = searchParams?.areaCodes ?? '';
  
  const config = getApiConfiguration();
  const indicatorApi = new IndicatorsApi(config);
  const data = await indicatorApi.getHealthDataForAnIndicator({
    indicatorId: Number(indicator),
    areaCodes: [areaCodes],
  });
  
  const filteredData = data
  .filter((item) => areaCodes === '' || item.areaCode === areaCodes)
  .map((item) => ({
    areaCode: item.areaCode,
    healthData: item.healthData,
  }));

  return (
    <Chart
      data={filteredData}
      indicator={indicator}
      indicatorsSelected={indicatorsSelected}
    />
  );
}
