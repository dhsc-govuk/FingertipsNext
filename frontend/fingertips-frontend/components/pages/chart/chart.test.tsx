import { render, screen } from '@testing-library/react';
import { Chart } from '@/components/pages/chart/index';
import { expect } from '@jest/globals';
import {WeatherForecast} from "@/generated-sources/api-client";

const mockData: WeatherForecast[] = [
  {
    date: new Date('2024-11-01T00:00:00.000Z'),
    temperatureC: -30,
    temperatureF: -21,
    summary: 'Freezing',
  },
  {
    date: new Date('2024-11-01T00:00:00.000Z'),
    temperatureC: 0,
    temperatureF: 32,
    summary: 'Bracing',
  },
];

test('snapshot test - should render the chart', () => {
  const container = render(<Chart data={mockData} />);

  expect(container.asFragment()).toMatchSnapshot();
});

test('should render the LineChart component with the correct props', () => {
  render(<Chart data={mockData} />);
  
  const lineChartTitle = screen.getByText('Line Chart');
  expect(lineChartTitle).toBeInTheDocument();
});

test('should render the PlainTable component', () => {
  render(<Chart data={mockData} />);
  
  const table = screen.getByRole('table');
  expect(table).toBeInTheDocument();
});
