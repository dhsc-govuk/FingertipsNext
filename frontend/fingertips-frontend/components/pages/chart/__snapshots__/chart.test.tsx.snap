// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test - should render the chart 1`] = `
<DocumentFragment>
  .c0 {
  color: #0b0c0c;
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 700;
  font-size: 32px;
  line-height: 1.09375;
  display: block;
  margin-top: 0;
  margin-bottom: 30px;
}

@media print {
  .c0 {
    color: #000;
  }
}

@media print {
  .c0 {
    font-size: 32px;
    line-height: 1.15;
  }
}

@media only screen and (min-width: 641px) {
  .c0 {
    font-size: 48px;
    line-height: 1.0416666666666667;
  }
}

@media only screen and (min-width: 641px) {
  .c0 {
    margin-bottom: 50px;
  }
}

<h1
    class="c0"
  >
    Line Chart
  </h1>
  @media print {

}

@media print {

}

@media only screen and (min-width: 641px) {

}

@media only screen and (min-width: 641px) {

}

<div>
    <div
      aria-hidden="false"
      aria-label="Weather Forecast. Highcharts interactive chart."
      data-highcharts-chart="0"
      data-testid="highcharts-react-component"
      role="region"
      style="overflow: hidden;"
    >
      <div
        aria-hidden="false"
        id="highcharts-screen-reader-region-before-0"
        style="position: relative;"
      >
        <div
          aria-hidden="false"
          style="position: absolute; width: 1px; height: 1px; overflow: hidden; white-space: nowrap; clip: rect(1px, 1px, 1px, 1px); margin-top: -3px; filter: alpha(opacity=1); opacity: 0.01;"
        >
          <h2>
            Weather Forecast
          </h2>
          <div>
            Line chart with 2 data points.
          </div>
          <div>
            A line chart showing temperature data over 6 months.
          </div>
          <div>
            The chart has 1 X axis displaying Date. 
          </div>
          <div>
            The chart has 1 Y axis displaying Temperature (°C). Data ranges from -30 to 0.
          </div>
        </div>
      </div>
      <div
        aria-hidden="false"
        class="highcharts-announcer-container"
        style="position: relative;"
      >
        <div
          aria-atomic="true"
          aria-hidden="false"
          aria-live="polite"
          style="position: absolute; width: 1px; height: 1px; overflow: hidden; white-space: nowrap; clip: rect(1px, 1px, 1px, 1px); margin-top: -3px; filter: alpha(opacity=1); opacity: 0.01;"
        />
        <div
          aria-atomic="true"
          aria-hidden="false"
          aria-live="assertive"
          style="position: absolute; width: 1px; height: 1px; overflow: hidden; white-space: nowrap; clip: rect(1px, 1px, 1px, 1px); margin-top: -3px; filter: alpha(opacity=1); opacity: 0.01;"
        />
        <div
          aria-atomic="true"
          aria-hidden="false"
          aria-live="polite"
          style="position: absolute; width: 1px; height: 1px; overflow: hidden; white-space: nowrap; clip: rect(1px, 1px, 1px, 1px); margin-top: -3px; filter: alpha(opacity=1); opacity: 0.01;"
        />
        <div
          aria-atomic="true"
          aria-hidden="false"
          aria-live="polite"
          style="position: absolute; width: 1px; height: 1px; overflow: hidden; white-space: nowrap; clip: rect(1px, 1px, 1px, 1px); margin-top: -3px; filter: alpha(opacity=1); opacity: 0.01;"
        />
      </div>
      <div
        aria-hidden="false"
        class="highcharts-container "
        dir="ltr"
        id="highcharts-mf5eu6t-0"
        style="position: relative; overflow: hidden; width: 600px; height: 400px; text-align: left; line-height: normal; z-index: 0; user-select: none; outline: none; padding: 0px;"
        tabindex="0"
      >
        <div
          aria-hidden="false"
          class="highcharts-a11y-proxy-container-before"
          style="top: 0px; left: 0px; white-space: nowrap; position: absolute;"
        />
        <svg
          aria-hidden="false"
          aria-label="Interactive chart"
          class="highcharts-root"
          height="400"
          style="font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", sans-serif; font-size: 1rem;"
          version="1.1"
          viewBox="0 0 600 400"
          width="600"
          xmlns="http://www.w3.org/2000/svg"
        >
          <desc
            aria-hidden="true"
          >
            Created with Highcharts 12.0.2
          </desc>
          <defs
            aria-hidden="true"
          >
            <filter
              id="highcharts-drop-shadow-0"
            >
              <fedropshadow
                dx="1"
                dy="1"
                flood-color="#000000"
                flood-opacity="0.75"
                stdDeviation="2.5"
              />
            </filter>
            <clippath
              id="highcharts-mf5eu6t-6-"
            >
              <rect
                fill="none"
                height="NaN"
                width="NaN"
                x="0"
                y="0"
              />
            </clippath>
            <clippath
              id="highcharts-mf5eu6t-7-"
            >
              <rect
                class="highcharts-animating"
                fill="none"
                height="NaNNaN"
                width="NaN"
                x="0"
                y="0"
              />
            </clippath>
            <clippath
              id="highcharts-mf5eu6t-8-"
            >
              <rect
                fill="none"
                height="NaN"
                width="99"
                x="-99"
                y="-99"
              />
            </clippath>
          </defs>
          <rect
            aria-hidden="true"
            class="highcharts-background"
            fill="#ffffff"
            filter="none"
            height="400"
            rx="0"
            ry="0"
            width="600"
            x="0"
            y="0"
          />
          <rect
            aria-hidden="true"
            class="highcharts-plot-background"
            fill="none"
            filter="none"
            height="NaN"
            width="NaN"
            x="NaN"
            y="10"
          />
          <g
            aria-hidden="true"
            class="highcharts-grid highcharts-xaxis-grid"
            data-z-index="1"
          >
            <path
              class="highcharts-grid-line"
              d="M 0 0"
              data-z-index="1"
              fill="none"
              opacity="1"
              stroke="#e6e6e6"
              stroke-dasharray="none"
              stroke-width="0"
            />
            <path
              class="highcharts-grid-line"
              d="M 0 0"
              data-z-index="1"
              fill="none"
              opacity="1"
              stroke="#e6e6e6"
              stroke-dasharray="none"
              stroke-width="0"
            />
            <path
              class="highcharts-grid-line"
              d="M 0 0"
              data-z-index="1"
              fill="none"
              opacity="1"
              stroke="#e6e6e6"
              stroke-dasharray="none"
              stroke-width="0"
            />
          </g>
          <g
            aria-hidden="true"
            class="highcharts-grid highcharts-yaxis-grid"
            data-z-index="1"
          />
          <rect
            aria-hidden="true"
            class="highcharts-plot-border"
            data-z-index="1"
            fill="none"
            height="0"
            stroke="#cccccc"
            stroke-width="0"
            width="0"
            x="0"
            y="10"
          />
          <g
            aria-hidden="true"
            class="highcharts-axis highcharts-xaxis"
            data-z-index="2"
          >
            <text
              class="highcharts-axis-title"
              data-z-index="7"
              style="color: rgb(102, 102, 102); font-size: 0.8em; fill: #666666;"
              text-anchor="middle"
              transform="translate(0,0)"
              x="NaN"
              y="NaN"
            >
              Date
            </text>
            <path
              class="highcharts-axis-line"
              d="M 0 0"
              data-z-index="7"
              fill="none"
              stroke="#333333"
              stroke-width="1"
            />
          </g>
          <g
            aria-hidden="true"
            class="highcharts-axis highcharts-yaxis"
            data-z-index="2"
          >
            <text
              class="highcharts-axis-title"
              data-z-index="7"
              style="color: rgb(102, 102, 102); font-size: 0.8em; fill: #666666;"
              text-anchor="middle"
              transform="translate(0,0) rotate(270 NaN NaN)"
              x="NaN"
              y="NaN"
            >
              Temperature (°C)
            </text>
            <path
              class="highcharts-axis-line"
              d="M 0 0"
              data-z-index="7"
              fill="none"
              stroke="#333333"
              stroke-width="0"
            />
          </g>
          <g
            aria-hidden="false"
            class="highcharts-series-group"
            data-z-index="3"
            filter="none"
          >
            <g
              aria-hidden="false"
              class="highcharts-series highcharts-series-0 highcharts-line-series highcharts-color-0"
              clip-path="url(#highcharts-mf5eu6t-7-)"
              data-z-index="0.1"
              opacity="1"
              transform="translate(NaN,10) scale(1 1)"
            />
            <g
              aria-hidden="true"
              class="highcharts-markers highcharts-series-0 highcharts-line-series highcharts-color-0"
              clip-path="url(#highcharts-mf5eu6t-8-)"
              data-z-index="0.1"
              opacity="1"
              transform="translate(NaN,10) scale(1 1)"
            />
          </g>
          <text
            aria-hidden="true"
            class="highcharts-title"
            data-z-index="4"
            style="font-size: 1.2em; color: rgb(51, 51, 51); font-weight: bold; fill: #333333;"
            text-anchor="middle"
            transform="translate(0,0) scale(1 1)"
            transform-origin="10 4"
            x="300"
            y="10"
          >
            Weather Forecast
          </text>
          <text
            aria-hidden="true"
            class="highcharts-subtitle"
            data-z-index="4"
            style="color: rgb(102, 102, 102); font-size: 0.8em; fill: #666666;"
            text-anchor="middle"
            transform="translate(0,0) scale(1 1)"
            transform-origin="10 3"
            x="300"
            y="10"
          />
          <text
            aria-hidden="true"
            class="highcharts-caption"
            data-z-index="4"
            style="color: rgb(102, 102, 102); font-size: 0.8em; fill: #666666;"
            text-anchor="start"
            transform="translate(0,0) scale(1 1)"
            transform-origin="10 3"
            x="10"
            y="387"
          />
          <g
            aria-hidden="true"
            class="highcharts-legend highcharts-no-tooltip"
            data-z-index="7"
            transform="translate(300,367)"
          >
            <rect
              class="highcharts-legend-box"
              fill="none"
              filter="none"
              rx="0"
              ry="0"
              stroke="#999999"
              stroke-width="0"
            />
            <g
              data-z-index="1"
            >
              <g>
                <g
                  class="highcharts-legend-item highcharts-line-series highcharts-color-0 highcharts-series-0"
                  data-z-index="1"
                  transform="translate(8,3)"
                >
                  <path
                    class="highcharts-graph"
                    d="M 1 4 L 15 4"
                    fill="none"
                    stroke="#2caffe"
                    stroke-linecap="round"
                    stroke-width="2"
                  />
                  <path
                    class="highcharts-point"
                    d="M 8 4 A 0 0 0 1 1 8 4 Z"
                    fill="#2caffe"
                    opacity="1"
                    stroke="#ffffff"
                    stroke-width="0"
                  />
                  <text
                    data-z-index="2"
                    style="color: rgb(51, 51, 51); cursor: pointer; font-size: 0.8em; text-decoration: none; fill: #333333;"
                    text-anchor="start"
                    x="21"
                    y="5"
                  >
                    Temperature (°C)
                  </text>
                </g>
              </g>
            </g>
          </g>
          <g
            aria-hidden="true"
            class="highcharts-axis-labels highcharts-xaxis-labels"
            data-z-index="7"
          >
            <text
              style="color: rgb(51, 51, 51); cursor: default; font-size: 0.8em; fill: #333333;"
              x="0"
            >
              x
            </text>
            <text
              style="color: rgb(51, 51, 51); cursor: default; font-size: 0.8em; fill: #333333;"
              text-anchor="middle"
              transform="translate(0,0)"
              x="0"
            >
              2024-11-01
            </text>
            <text
              style="color: rgb(51, 51, 51); cursor: default; font-size: 0.8em; fill: #333333;"
              text-anchor="middle"
              transform="translate(0,0)"
              x="0"
            >
              2024-11-01
            </text>
          </g>
          <g
            aria-hidden="true"
            class="highcharts-axis-labels highcharts-yaxis-labels"
            data-z-index="7"
          />
          <text
            aria-hidden="false"
            aria-label="Chart credits: Highcharts.com"
            class="highcharts-credits"
            data-z-index="8"
            style="cursor: pointer; color: rgb(153, 153, 153); font-size: 0.6em; fill: #999999;"
            text-anchor="end"
            x="590"
            y="395"
          >
            Highcharts.com
          </text>
        </svg>
        <div
          aria-hidden="false"
          class="highcharts-a11y-proxy-container-after"
          style="top: 0px; left: 0px; white-space: nowrap; position: absolute;"
        >
          <div
            class="highcharts-a11y-proxy-group highcharts-a11y-proxy-group-zoom"
          />
          <div
            aria-label="Toggle series visibility, Weather Forecast"
            class="highcharts-a11y-proxy-group highcharts-a11y-proxy-group-legend"
            role="region"
          >
            <ul
              role="list"
            >
              <li
                style="list-style: none;"
              >
                <button
                  aria-label="Show Temperature (°C)"
                  aria-pressed="true"
                  class="highcharts-a11y-proxy-element"
                  style="border-width: 0px; background-color: transparent; cursor: pointer; outline: none; opacity: 0.001; filter: alpha(opacity=1); z-index: 999; overflow: hidden; padding: 0px; margin: 0px; display: block; position: absolute; width: 1px; height: 1px; left: 0px; top: 0px;"
                  tabindex="-1"
                />
              </li>
            </ul>
          </div>
          <div
            class="highcharts-a11y-proxy-group highcharts-a11y-proxy-group-chartMenu"
          />
        </div>
      </div>
      <div
        aria-hidden="false"
        id="highcharts-screen-reader-region-after-0"
        style="position: relative;"
      >
        <div
          aria-hidden="false"
          style="position: absolute; width: 1px; height: 1px; overflow: hidden; white-space: nowrap; clip: rect(1px, 1px, 1px, 1px); margin-top: -3px; filter: alpha(opacity=1); opacity: 0.01;"
        >
          <div
            aria-hidden="false"
            class="highcharts-exit-anchor"
            id="highcharts-end-of-chart-marker-0"
            tabindex="0"
          >
            End of interactive chart.
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
